app/com/arpnetworking/kairos/client/KairosDbClient.java

```diff
@@ -0,0 +1,182 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import net.sf.oval.constraint.NotNull;
+import scala.compat.java8.FutureConverters;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Client for accessing KairosDB APIs.
+ *
+ * @author Brandon Arp (brandon dot arp at smartsheet dot com)
+ */
+public final class KairosDbClient {
+    /**
+     * Executes a query for datapoints from  KairosDB.
+     *
+     * @param query the query
+     * @return the response
+     */
+    public CompletionStage<KairosQuery> queryMetrics(final MetricsQuery query) {
+        try {
+            final HttpRequest request = HttpRequest.POST(createUri(METRICS_QUERY_PATH))
+                    .withEntity(ContentTypes.APPLICATION_JSON, _mapper.writeValueAsString(query));
+            return fireRequest(request, MetricsQueryResponse.class)
+                    .thenApply(response -> new KairosQuery.Builder().setRequest(query).setResponse(response).build());
+        } catch (final JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Queries KairosDB for metric names.
+     *
+     * @return the response
+     */
+    public CompletionStage<KairosMetricNamesQueryResponse> queryMetricNames() {
+        final HttpRequest request = HttpRequest.GET(createUri(METRICS_NAMES_PATH));
+        return fireRequest(request, KairosMetricNamesQueryResponse.class);
+    }
+
+    private <T> CompletionStage<T> fireRequest(final HttpRequest request, final Class<T> responseType) {
+        return _http.singleRequest(request.addHeader(AcceptEncoding.create(HttpEncodings.GZIP)), _materializer)
+                .thenCompose(httpResponse -> {
+                    final HttpEncoding encoding = httpResponse.encoding();
+                    final Coder flow;
+                    if (encoding == HttpEncodings.GZIP) {
+                        flow = Gzip$.MODULE$;
+                    } else if (encoding == HttpEncodings.DEFLATE) {
+                        flow = Deflate$.MODULE$;
+                    } else {
+                        flow = NoCoding$.MODULE$;
+                    }
+                    return httpResponse.entity()
+                            .toStrict(READ_TIMEOUT.toMillis(), _materializer)
+                            .thenCompose(strict -> FutureConverters.toJava(flow.decode(strict.getData(), _materializer)));
+                })
+                .thenApply(body -> {
+                    try {
+                        return _mapper.readValue(body.utf8String(), responseType);
+                    } catch (final IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+    }
+
+    private String createUri(final URI relativePath) {
```

Comment:
Should return a `URI`; otherwise rename.

Reply:
Fixed.